The Great Swift 4 Refactor

AttilaTheFun opened the pull request:
This PR addresses https://github.com/AttilaTheFun/SwaggerParser/issues/3 and refactors the library to remove the ObjectMapper dependency in favor of the native Codable protocol. It also includes some other smaller improvements like a nicer Version API and it leverages some of the dictionary methods built in to the standard library in Swift 4.

n8chur commented on Sources/Enums.swift:
```diff
@@ -1,19 +1,26 @@
+    case header = "header"
+    case path = "path"
+    case formData = "formData"
+    case body = "body"
```
Is it necessary to set these explicitly to their string values if the enum is `: String`? Why not omit the string value if the value's name matches the string representation?

n8chur commented on Sources/Enums.swift:
```diff
@@ -79,24 +86,35 @@ public enum DataType: String {
+            guard let typeString = try? values.decode(String.self, forKey: .type),
+                let type = DataType(rawValue: typeString) else
+            {
+                throw DecodingError("Unknown data type")
```
Should the `typeString` be included in the `DecodingError`?

n8chur commented on Sources/FixedParameterFields.swift:
```diff
@@ -29,24 +20,35 @@ public struct FixedParameterFields {
+        self.description = try? values.decode(String.self, forKey: .description)
+        self.required = (try values.decodeIfPresent(Bool.self, forKey: .required)) ?? false
+        self.example = try? values.decode(String.self, forKey: .example)
+        // TODO: Decode example as any.
```
What's preventing this from being done now?

AttilaTheFun replied:
I had to write extensions on Keyed(Encoding|Decoding)Container to allow encoding of Any types. Normally the encode/decode functions only take either a generic param conforming to the respective encoding/decoding protocol, or one of a limited set of predefined primitive types. (E.g. String, Int64 etc). I've done this now, though, if you want to take a look.

n8chur commented on Sources/Metadata.swift:
```diff
@@ -24,32 +23,58 @@ public struct Metadata {
     let description: String?
     let defaultValue: Any?
-    let enumeratedValues: [Any?]?
+    let enumeratedValues: [Any]?
```
Is nil not a valid value?

n8chur commented on Sources/Metadata.swift:
```diff
@@ -24,32 +23,58 @@ public struct Metadata {
+        // enumeratedValues
+        try container.encode(self.nullable, forKey: .nullable)
+        // example
+        // TODO: Encode any values.
```
Will this be completed in this PR?

n8chur commented on Sources/ObjectSchema.swift:
```diff
@@ -10,64 +12,43 @@ public struct ObjectSchema {
     /// that is, properties whose names are not listed in the properties keyword. 
     /// By default any additional properties are allowed.
-    /// The additionalProperties may be either a boolean or a schema.
+    /// The additionalProperties may bdde either a boolean or a schema.
```
Typo

n8chur commented on Sources/Operation.swift:
```diff
@@ -44,52 +43,83 @@ public struct Operation {
+        try container.encode(self.parameters, forKey: .parameters)
+        var allResponses = [String: Reference<ResponseBuilder>]()
+        allResponses["default"] = self.defaultResponse
+        self.responses.forEach { allResponses[$0.description] = $1 }
```
Is `<Int>.description` preferable over `String(<Int>)`?

n8chur commented on Sources/Version.swift:
```diff
@@ -1,39 +1,131 @@
+                return int
+            }
+
+            throw DecodingError("Unable to parse version component into integer")
```
Should we add the `component` to this error string to assist in debugging?

n8chur commented on Sources/Version.swift:
```diff
@@ -1,39 +1,131 @@
+            return Version.subversion(component, subversion)
+        }
+
+        self = version
```
Why not set self directly above?:
```
self = integerComponents.dropLast().reversed().reduce(Version.version(lastComponent))
...
```

n8chur commented on Sources/Version.swift:
```diff
@@ -1,39 +1,131 @@
+
+extension Version: Comparable {
+    public static func <(left: Version, right: Version) -> Bool {
+        let leftArray = left.map { $0.value }
```
Should there be a `components` property on `Version`?

AttilaTheFun replied:
Sure

n8chur commented on Sources/Version.swift:
```diff
@@ -1,39 +1,131 @@
+    public var minor: UInt? { return self.subversion?.value }
+    public var patch: UInt? { return self.subversion?.subversion?.value }
+
+    public init(from decoder: Decoder) throws {
```
Should we add a convenience initializer that takes a `String` (that this initializer calls)? I imagine that could be useful for version comparisons by the consumer.

AttilaTheFun replied:
Sure

AttilaTheFun commented:
@n8chur Okay, I think I've addressed all of your comments. @yonaskolb would you like to take a look? I'm planning on merging this and cutting a release after Xcode 9 GM goes live next week.
